wordpress-mobile/WordPress-iOS
# Posts & Pages: Autosizing and Readability

While I originally set out to adapt Posts and Pages away from fixed widths, it ultimately brought me to this glorious refactoring for autosizing cells and following iOS imposed margins.

### Overview
- Cleans up the layout for Pages and Posts.
- Autosizes cells via `UITableViewAutomaticDimension`.
- Adopts readable margins for iPad and iPhone 6+ landscape (Reg Width and Any Height)
- Updates the Pages design just a _bit_.

For now we'll only follow readable margins on iPad and iPhone 6+ landscape. This conforms to the size classes of _Regular Width_ and _Any Height_. We'll leave default iPhone as is for now, as the margins there are technically hand-designed.

### Estimated Height Cacheing
With the autosizing cells, I absolutely could not find a way to keep cells from _bouncing_ their heights after any reload animations. Even if the cell's intrinsic size doesn't change, `UITableView` seems to start a cell's reload animation by key-framing the beginning height as whatever the estimated height is. So unless the estimated height is _precisely_ the layout of the cell, the animation will appear to bounce after firing off the reload calls.

To solve this, of course, we're cacheing the last known frame heights of the cell's based off of `willDisplayCell:`. This cached height is then used as the explicit estimated height for the given indexPath. This doesn't seem quite _auto_, but I can't find a better solution at this time.

### Pages Design
While adapting Pages for readability, the design had to be updated in accordance with the rest of the app and our iOS consistency patterns.

-

### Testing, Posts:

1. Open Posts on iPad, ensure the Post's content views are following readable margins on portrait and landscape orientations.
2. Open Posts via iPad multitasking, ensure the Post's content views are defaulting to the compact sized class margins, as also seen on iPhone. Ensure the margins transition as expected.
3. Open Posts on iPhone, ensure the Post's content views are defaulting to the compact sized class margins, as originally expected.
4. Open Posts on iPhone 6+ landscape, ensure the Post's content views are following readable margins on landscape orientation.
5. Transition from Posts on iPhone 6+ landscape to portrait, ensure the margins transition as expected.
6. Move a Post to Trash, ensure the Restore cell is loaded as expected and vice versa.
7. Ensure Post cell heights are correct and post content is only truncated as expected. (Compare to Store build)

### Testing, Pages:

1. Open Pages on iPhone and iPad, ensure the cell contents are following readable margins.
2. Move a Page to Trash, ensure the Restore cell is loaded as expected and vice versa.
3. Ensure Page cell heights are correct and also handle long, multi-line page titles.

Needs review: @frosty can I interest you in taking a dive into autosizing post cells? 😅
